java-checks/src/main/java/org/sonar/java/checks/tests/MockitoAnnotatedObjectsShouldBeInitializedCheck.java

```diff
@@ -0,0 +1,191 @@
+    "org.junit.Before",
+    "org.junit.jupiter.api.BeforeEach"
+  );
+
+  private static final String RULE_ANNOTATION = "org.junit.Rule";
+
+  private static final MethodMatchers MOCKITO_JUNIT_RULE = MethodMatchers.create()
+    .ofAnyType()
+    .names("rule")
+    .addWithoutParametersMatcher()
+    .build();
+
+  private static final String MESSAGE = "Initialize mocks before using them.";
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return Collections.singletonList(Tree.Kind.CLASS);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    ClassTree testClass = (ClassTree) tree;
+    List<VariableTree> mocksToInitialize = testClass.members().stream()
+      .filter(MockitoAnnotatedObjectsShouldBeInitializedCheck::isFieldWithTargetAnnotation)
+      .map(VariableTree.class::cast)
+      .collect(Collectors.toList());
+    if (!mocksToInitialize.isEmpty() && !mocksAreProperlyInitialized(testClass)) {
+      AnnotationTree firstAnnotation = mocksToInitialize.get(0).modifiers().annotations().get(0);
+      reportIssue(firstAnnotation, MESSAGE);
+    }
+  }
+
+  private static boolean isFieldWithTargetAnnotation(Tree tree) {
+    if (!tree.is(Tree.Kind.VARIABLE)) {
+      return false;
+    }
+    VariableTree field = (VariableTree) tree;
+    SymbolMetadata metadata = field.symbol().metadata();
+    return TARGET_ANNOTATIONS.stream().anyMatch(metadata::isAnnotatedWith);
+  }
+
+  private static boolean mocksAreProperlyInitialized(ClassTree testClass) {
+    return isClassProperlyAnnotated(testClass) ||
+      isMockitoJUnitRuleInvoked(testClass) ||
+      areMocksInitializedInSetup(testClass);
+  }
+
+  public static boolean isClassProperlyAnnotated(ClassTree clazz) {
+    SymbolMetadata metadata = clazz.symbol().metadata();
+    return EXPECTED_CLASS_ANNOTATIONS.stream().anyMatch(metadata::isAnnotatedWith);
+  }
+
+  private static boolean isMockitoJUnitRuleInvoked(ClassTree clazz) {
+    List<VariableTree> collected = clazz.members().stream()
+      .filter(member -> member.is(Tree.Kind.VARIABLE))
+      .map(VariableTree.class::cast)
+      .collect(Collectors.toList());
+    for (VariableTree field : collected) {
+      if (field.type().symbolType().is("org.mockito.junit.MockitoRule")) {
+        ExpressionTree initializer = field.initializer();
+        if (initializer != null && initializer.is(Tree.Kind.METHOD_INVOCATION) &&
+          MOCKITO_JUNIT_RULE.matches((MethodInvocationTree) initializer) &&
+          field.symbol().metadata().isAnnotatedWith(RULE_ANNOTATION)) {
+          return true;
+        }
+      }
+    }
+    return false;
+  }
+
+  private static boolean areMocksInitializedInSetup(ClassTree clazz) {
+    List<MethodTree> methods = new ArrayList<>();
+    Optional<ClassTree> parent = Optional.of(clazz);
+    while (parent.isPresent()) {
+      ClassTree tree = parent.get();
+      methods.addAll(getSetupMethods(tree));
+      parent = getParentClass(tree);
```

Comment:
In our unit test, it's a corner case where from a ClassTree we can access to the super ClassTree because all code is in the same file.
Usually, from a ClassTree we can only access to the semantic of the superclass, but not the AST.
With the semantic, it's not possible to apply SetupMethodVisitor.
So, I would just not support the case where the ClassTree have superClass() as a limitation of the rule, with a unit test that show a false-negative, and remove this getParentClass logic.